src/directory/s3_directory.rs

```diff
@@ -0,0 +1,285 @@
+use std::fmt;
+use std::io;
+use std::path::{Path, PathBuf};
+use std::result;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::sync::RwLock;
+use rusoto_core::{DefaultCredentialsProvider, Region, default_tls_client};
+use rusoto_s3::{S3, S3Client, HeadBucketRequest, GetObjectRequest};
+
+fn get_client(region: Region) -> Result<Box<S3>, Box<Error>> {
+    // TODO: handle missing creds
+    let client = default_tls_client()?;
+    let provider = DefaultCredentialsProvider::new()?;
+
+    Ok(Box::new(S3Client::new(client, provider, region)))
+}
+
+#[derive(Clone)]
+struct InnerDirectory {
+    bucket: String,
+    cache: Arc<RwLock<HashMap<PathBuf, Arc<Vec<u8>>>>>,
+}
+
+impl InnerDirectory {
+    fn new(bucket: String) -> InnerDirectory {
+        InnerDirectory {
+            cache: Arc::new(RwLock::new(HashMap::new())),
+            bucket,
+        }
+    }
+
+    fn load_missing_key(&self, client: &S3, path: &Path) -> Result<(), OpenReadError> {
+        let mut map = self.cache.write().map_err(|_| {
+            let msg = format!(
+                "Failed to acquire write lock for the \
+                                            directory when trying to read {:?}",
+                path
+            );
+            let io_err = make_io_err(msg);
+            OpenReadError::IOError(IOError::with_path(path.to_owned(), io_err))
+        })?;
+
+        // TODO: this is comical and I'm more than likely over thinking it
+        let key = path.as_os_str().to_os_string().into_string().map_err(|_| {
+            let msg = format!("Could not build key path");
+            let io_err = make_io_err(msg);
+            OpenReadError::IOError(IOError::with_path(path.to_owned(), io_err))
+        })?;
+
+        let obj = client
+            .get_object(&GetObjectRequest {
+                bucket: self.bucket.clone(),
+                key,
+                ..Default::default()
+            })
+            .map_err(|_| {
+                let msg = format!("No key found for {:?}", path);
+                let io_err = make_io_err(msg);
+                OpenReadError::IOError(IOError::with_path(path.to_owned(), io_err))
+            })?;
+
+        map.insert(PathBuf::from(path), Arc::new(obj.body.unwrap()));
+
+        Ok(())
+    }
+
+    fn open_read(&self, path: &Path, client: &S3) -> result::Result<ReadOnlySource, OpenReadError> {
+        debug!("Open Read {:?}", path);
+
+        let cache = self.cache.read().map_err(|_| {
+            let msg = format!(
+                "Failed to acquire read lock for the \
+                                            directory when trying to read {:?}",
+                path
+            );
+            let io_err = make_io_err(msg);
+            OpenReadError::IOError(IOError::with_path(path.to_owned(), io_err))
+        })?;
+
+        if !cache.contains_key(path) {
+            self.load_missing_key(client, path)?;
+        }
+
+        //TODO: map_err
+        let data = cache.get(path).unwrap();
```

Comment:
Supposedly it is not possible, because if it was the case, load_missing_key would have returned an Error right? 
But this assumes load_missing_key has been implemented correctly of course... And having to deal with the None branch feels a bit clumsy...

What if `load_missing_key` was called `fetch` and was returning the data? You could then use the entry interface of the map in this method.